packages/evm/src/eof/container.ts

```diff
@@ -0,0 +1,461 @@
+        }
+        if (outputs !== 0x80) {
+          validationError(EOFError.Code0Outputs)
+        }
+      }
+      if (inputs > INPUTS_MAX) {
+        validationError(EOFError.MaxInputs, i, inputs)
+      }
+      if (outputs > OUTPUTS_MAX) {
+        validationError(EOFError.MaxOutputs, i, outputs)
+      }
+      if (maxStackHeight > MAX_STACK_HEIGHT) {
+        validationError(EOFError.MaxStackHeightLimit, i, maxStackHeight)
+      }
+      typeSections.push({
+        inputs,
+        outputs,
+        maxStackHeight,
+      })
+    }
+    // Read each code section
+    const codeStartPtr = stream.getPtr()
+    const codes = []
+    for (const [i, codeSize] of header.codeSizes.entries()) {
+      try {
+        const code = stream.readBytes(codeSize)
+        codes.push(code)
+      } catch {
+        validationError(EOFError.CodeSection, i)
+      }
+    }
+    // Write the entire code section to the entireCodeSection
+    const entireCodeSection = buf.slice(codeStartPtr, stream.getPtr())
+
+    // Read all raw subcontainers and push those to the containers array
+    const containers = []
+    for (const [i, containerSize] of header.containerSizes.entries()) {
+      try {
+        const container = stream.readBytes(containerSize)
+        containers.push(container)
+      } catch {
+        validationError(EOFError.ContainerSection, i)
+      }
+    }
+
+    // Data section of the body
+    // Note: for EOF containers in Initmode (these are Subcontainers) it is allowed
+    // to have a data section of size lower than what is written in the header
+    // For details, see "Data section lifecycle" of EIP 7620
+    let dataSection: Uint8Array
+
+    // Edge case: deployment code validation
+    if (eofMode !== EOFContainerMode.Initmode && !dataSectionAllowedSmaller) {
+      dataSection = stream.readBytes(header.dataSize, EOFError.DataSection)
+
+      if (eofMode === EOFContainerMode.Default) {
+        if (!stream.isAtEnd()) {
+          // If there are dangling bytes in default container mode, this is invalid
+          validationError(EOFError.DanglingBytes)
+        }
+      } else {
+        // Tx init mode: the remaining bytes (if any) are used as CALLDATA in the EVM, in case of a Tx init
+        this.txCallData = stream.readRemainder()
+      }
+    } else {
+      dataSection = stream.readRemainder()
+    }
+
+    // Write all data to the object
+    this.typeSections = typeSections
+    this.codeSections = codes
+    this.containerSections = containers
+    this.entireCode = entireCodeSection
+    this.dataSection = dataSection
+    this.buffer = buf
+  }
+  sections() {
+    return [this.typeSections, this.codeSections, this.dataSection]
+  }
+  size() {
+    return {
+      typeSize: this.typeSections.length,
+      codeSize: this.codeSections.length,
+      dataSize: this.dataSection.length,
+    }
+  }
+  sectionSizes() {
+    return [
+      this.typeSections.map(() => 4),
```

Comment:
Feels like a magic number here so would be nice to either add the below comment (if correct) or replace `4` with a constant that's declared elsewhere (in `constants.ts`?)
```suggestion
      // A type section size is always 4 bytes long
      this.typeSections.map(() => 4),
```